Data Preprocessing:

Loading the Dataset:

In [3]:
import pandas as pd

In [4]:
df=pd.read_csv("C:\\DataScience25\\official_assignments\\Recommendation System\\Recommendation System\\anime.csv")

In [5]:
df

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

In [6]:
df.shape

(12294, 7)

In [7]:
df.columns

Index(['anime_id', 'name', 'genre', 'type', 'episodes', 'rating', 'members'], dtype='object')

Checking for missing values:

In [9]:
df.isnull().sum()

anime_id      0
name          0
genre        62
type         25
episodes      0
rating      230
members       0
dtype: int64

Handling Missing Values:

In [11]:
df['genre']=df['genre'].fillna(df['genre'].mode()[0])

In [12]:
df['type']=df['type'].fillna(df['type'].mode()[0])

In [13]:
df['rating']=df['rating'].fillna(df['rating'].median())

In [14]:
df.isnull().sum()

anime_id    0
name        0
genre       0
type        0
episodes    0
rating      0
members     0
dtype: int64

Handled the missing values of df by impututing Genre and Type with mode as they are categorical features and for rating imputed using median as it is a numerical feature.

Handling categorical features:

In [17]:
df.select_dtypes(include='object').columns

Index(['name', 'genre', 'type', 'episodes'], dtype='object')

For genre, we are using MultiLabelBinarizer as there are multiple comma separated labels in a single column

In [19]:
from sklearn.preprocessing import MultiLabelBinarizer
mlb=MultiLabelBinarizer()
genre_binarized=mlb.fit_transform(df['genre'].str.split(', '))
genre_df = pd.DataFrame(genre_binarized, columns=mlb.classes_)

In [20]:
df=df.drop('genre',axis=1)

Concating genre_df with the original df:

In [22]:
df = pd.concat([df, genre_df], axis=1)

In [23]:
df=pd.get_dummies(df,columns=['type'])

Converting episode to numeric:

In [25]:
df['episodes'] = pd.to_numeric(df['episodes'], errors='coerce')

Statistical Summary of the Numerical Columns:

In [27]:
df.describe()

anime_id      episodes        rating       members        Action  \
count  12294.000000  11954.000000  12294.000000  1.229400e+04  12294.000000   
mean   14058.221653     12.382550      6.475700  1.807134e+04      0.231414   
std    11455.294701     46.865352      1.017179  5.482068e+04      0.421753   
min        1.000000      1.000000      1.670000  5.000000e+00      0.000000   
25%     3484.250000      1.000000      5.900000  2.250000e+02      0.000000   
50%    10260.500000      2.000000      6.570000  1.550000e+03      0.000000   
75%    24794.500000     12.000000      7.170000  9.437000e+03      0.000000   
max    34527.000000   1818.000000     10.000000  1.013917e+06      1.000000   

          Adventure          Cars        Comedy      Dementia        Demons  \
count  12294.000000  12294.000000  12294.000000  12294.000000  12294.000000   
mean       0.190987      0.005857      0.377827      0.019522      0.023914   
std        0.393095      0.076307      0.484864      0.138355      0.152788   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      1.000000      0.000000      0.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

       ...    Shounen Ai  Slice of Life         Space        Sports  \
count  ...  12294.000000   12294.000000  12294.000000  12294.000000   
mean   ...      0.005287       0.099235      0.030991      0.044168   
std    ...      0.072523       0.298990      0.173300      0.205476   
min    ...      0.000000       0.000000      0.000000      0.000000   
25%    ...      0.000000       0.000000      0.000000      0.000000   
50%    ...      0.000000       0.000000      0.000000      0.000000   
75%    ...      0.000000       0.000000      0.000000      0.000000   
max    ...      1.000000       1.000000      1.000000      1.000000   

        Super Power  Supernatural      Thriller       Vampire          Yaoi  \
count  12294.000000  12294.000000  12294.000000  12294.000000  12294.000000   
mean       0.037823      0.084350      0.007077      0.008297      0.003172   
std        0.190777      0.277923      0.083828      0.090711      0.056236   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

               Yuri  
count  12294.000000  
mean       0.003416  
std        0.058352  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000  

[8 rows x 47 columns]

- Episode:      
Mean number of episodes are ~12.38.                
High standard deviations and the max value indicates the presence of outliers.

- Rating:         
Ratings are ranging from 1.67 and 10.

- Members:                          
Mean number of members per anime are ~18,000.                   
Max number of members are ~1 million.

- Genre Features:                           
Action is present in 23% of anime.                     
Adventure in 19% of the anime.                     
Most frequent is Comdey which is presnt in 37% of the anime.

Feature Extraction:


Decide on the features that will be used for computing similarity (e.g., genres, user ratings):


Using the genre columns for similarity as they directly describe the content or theme of the Anime.                     
Cosine similarity works best with Binary Vectors.

Recommendation System(Content-Based Filtering):


Design a function to recommend anime based on cosine similarity.


Given a target anime, recommend a list of similar anime based on cosine similarity scores.


Experiment with different threshold values for similarity scores to adjust the recommendation list size.


In [36]:
def recommend_anime_threshold(anime_name, df, threshold=0.5):
    
    #anime_name: The name of the reference anime
    #df: DataFrame containing 'name' and genre binary columns
    #threshold: Similarity score threshold (0 to 1)

    from sklearn.metrics.pairwise import cosine_similarity

    # Check if anime exists
    if anime_name not in df['name'].values:
        return f"'{anime_name}' not found in dataset."

    # Genre matrix
    genre_df = df.loc[:, 'Action':'Yuri']  # adjust genre column range

    # Cosine similarity matrix
    cosine_sim = cosine_similarity(genre_df)

    # Index of target anime
    idx = df[df['name'] == anime_name].index[0]

    # Similarity scores with all other anime
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Filter using threshold (excluding the anime itself)
    filtered = [(i, score) for i, score in sim_scores if score >= threshold and i != idx]

    # Sort by similarity
    filtered_sorted = sorted(filtered, key=lambda x: x[1], reverse=True)

    # Get anime names
    recommendations = [(df.iloc[i]['name'], score) for i, score in filtered_sorted]

    return pd.DataFrame(recommendations, columns=['Recommended Anime', 'Similarity Score'])


In [38]:
recommend_anime_threshold("Kimi no Na wa.", df)

Recommended Anime  Similarity Score
0                          Wind: A Breath of Heart OVA          1.000000
1                         Wind: A Breath of Heart (TV)          1.000000
2                Aura: Maryuuin Kouga Saigo no Tatakai          0.894427
3                        Kokoro ga Sakebitagatterunda.          0.866025
4    Clannad: After Story - Mou Hitotsu no Sekai, K...          0.866025
..                                                 ...               ...
824                                          Kakyuusei          0.500000
825                                Ashita no Yukinojou          0.500000
826                Doukyuusei (OVA): Natsu no Owari ni          0.500000
827                             Stratosphera no Yousei          0.500000
828                        Gloria: Kindan no Ketsuzoku          0.500000

[829 rows x 2 columns]

In [40]:
recommend_anime_threshold("Naruto", df, threshold=0.7)


Recommended Anime  Similarity Score
0                             Boruto: Naruto the Movie          1.000000
1                                   Naruto: Shippuuden          1.000000
2    Boruto: Naruto the Movie - Naruto ga Hokage ni...          1.000000
3                                          Naruto x UT          1.000000
4          Naruto: Shippuuden Movie 4 - The Lost Tower          1.000000
..                                                 ...               ...
113                                          Tokyo ESP          0.730297
114                                          Gatchaman          0.730297
115                                      Ninku (Movie)          0.730297
116                                     Baoh Raihousha          0.730297
117                                       Taboo Tattoo          0.730297

[118 rows x 2 columns]

Evaluation:


Since the recommendation system is based on content similarity using features like genre, rating, and member count, there was no training involved. Therefore, splitting the dataset into training and testing sets is not required.

Moreover, since there is no user-item interaction or feedback data available, it is not possible to calculate standard evaluation metrics like precision, recall, or F1-score.

However, the performance of the system can be improved in the following ways:

- Incorporating user viewing or rating history to enable collaborative filtering

- Adding more features such as release year, studio, or tags

- Normalizing or weighting features like rating and members to balance their influence on similarity scores

Interview Questions:

Q1: Can you explain the difference between user-based and item-based collaborative filtering?

- User-Based Collaborative Filtering recommends items to a user based on the preferences of similar users.

It identifies users who have similar tastes and recommends what they liked.

Example: If User A and User B both like Anime X, and User A also likes Anime Y, then Anime Y is recommended to User B.

- Item-Based Collaborative Filtering recommends items that are similar to those the user already liked.

It finds items with similar rating patterns across users.

Example: If many users who liked Anime X also liked Anime Y, and a user watched Anime X, then Anime Y is recommended.

 Q2: What is collaborative filtering, and how does it work?

Collaborative filtering is a recommendation technique that suggests items to a user based on the preferences of similar users or similar items. It assumes that users with similar tastes will prefer similar things in the future.

It works by using a user-item interaction matrix — like user ratings, watch history, or likes — and calculates similarity between users or items.

This method needs actual user behavior data, not just item features.